lib/astr.3
.Sh DESCRIPTION
The
.Nm
string library provides a high-level interface to dynamically
allocated null-terminated C strings.
.Pp
The library provides two types:
.Fa astr
and
.Fa castr .
The type
.Fa astr
is a modifiable string, while
.Fa castr
is a const (read-only) string.
Strings with type
.Fa astr
can be implicitly casted to
.Fa castr ,
while the (explicitly) inverse operation should be avoided.
.Pp
Most of the user code will use the
.Fa astr
type, while
.Fa castr
will be used as function arguments to signal that they
will not be modified by the function.
.Pp
The
.Fn astr_cmp
function compares the argument strings and returns an integer less than,
equal to, or greater than zero if
.Fa s1
is found, respectively, to be less than, to match, or be greater than
.Fa s2 .
.Pp
The
.Fn astr_assign ,
and
.Fn astr_assign_cstr
functions assign the contents of the argument string or character
to the string
.Fa as ,
overwritting any previously existing text.
.Pp
The
.Fn astr_append ,
.Fn astr_append_cstr ,
and
.Fn astr_append_char
functions append the contents of the argument string or character
at the end of the string
.Fa as .
.Pp
The
.Fn astr_substr
function returns a new allocated substring extracted from the argument string
.Fa as ,
starting from position
.Fa pos
up to
.Fa size
characters.
Negative
.Fa pos
values mean
starting counting from the end of the string, instead of the beginning.
.Pp
The
.Fn astr_last_char
function returns the last character in the string.
.Pp
The
.Fn astr_find
and
.Fn astr_find_cstr
functions find the first occurrence of the argument string or character
into the argument
.Fa as ,
returning the position starting from the beginning of the string.
.Pp
The
.Fn astr_rfind
and
.Fn astr_rfind_cstr
functions find the first occurrence of the argument string or character
into the argument
.Fa as ,
returning the position starting from the end of the string.
.Pp
The
.Fn astr_replace ,
and
.Fn astr_replace_cstr
functions replace up to
.Fa size
characters of the argument string
.Fa as ,
starting from the position
.Fa pos ,
with the argument string or character.
.Pp
The
.Fn astr_fgets
function reads a string from the file
.Fa f
and stores it into the string
.Fa as .
The trailing newline is removed from the string.
The
.Fn astr_fgets
function returns the argument
.Fa as
string, and
.Fa NULL
when end of file occurs while no characters have been read.
.Pp
The
.Fn astr_fmt
and
.Fn astr_vfmt
functions write formatted text into the argument string
.Fa as .
Previously existing text will be overwritten.
.Pp
The
.Fn astr_afmt
and
.Fn astr_vafmt
functions append formatted text to the argument string
.Fa as .
.Pp
Where not otherwise specified, the
.Nm
functions return the first argument string, usually named
.Fa as
in the function prototype.
.Sh IMPLEMENTATION
Internally, each
.Nm
entry stores three variables: a buffer that contains
the C string, the buffer size and the size of the string.
.Pp
At each modification that enlarges the string, if the buffer is not big
enough to contain the modified string, the buffer is reallocated with
.Fa realloc .
.Sh EXAMPLES
Create and concatenate two strings:
.Bd -literal -offset indent
astr s1, s2;
s1 = astr_new();
s2 = astr_new();
astr_assign_cstr(s1, "string 1");
astr_assign_cstr(s2, "string 2");
astr_append_char(s1, ' ');
astr_append(s1, s2);
printf("s1: '%s', s2: '%s'\\n", astr_cstr(s1), astr_cstr(s2));
astr_delete(s1);
astr_delete(s2);
.Ed
.Pp
Replace occurrences into a string:
.Bd -literal -offset indent
astr s;
int i;
s = astr_new();
astr_assign_cstr(s, "buffer: this is a buffer");
while ((i = astr_find_cstr(s, "buffer")) >= 0)
	astr_replace_cstr(s, i, 6, "astr string");
printf("s: '%s'\\n", astr_cstr(s));
astr_delete(s);
.Ed
.Pp
Filtering files:
.Bd -literal -offset indent
astr s;
int i;
s = astr_new();
while (astr_fgets(s, stdin) != NULL) {
	while ((i = astr_find_cstr(s, "vi")) >= 0)
		astr_replace_cstr(s, i, 2, "emacs");
	astr_fputs(s, stdout);
	putc('\\n', stdout);
}
astr_delete(s);
.Ed
.Pp
Miscellaneous operations on strings:
.Bd -literal -offset indent
astr s;
s = astr_new();
astr_assign_cstr(s, "world");
astr_prepend_cstr(s, "hello");
astr_append_char(s, '!');
printf("s: '%s'\\n", astr_cstr(s));
astr_delete(s);
.Ed


NEWS
Changes between Zile 1.6.1 - Zile 1.7 BETA 2

Note: 1.7 BETA 2 was produced from 1.6.1, not 1.6.2, but all the same
fixes were applied.

* Changed configure script so no syntax colouring modes are built by
  default.
* Implemented Java Mode.  Contributed by Marius Nita <marius@cs.pdx.edu>.
* Implemented C# Mode.  Contributed by Josef 'Jupp' Schugt <jupp@gmx.de>.
* Implemented Mail Mode (plus the five variables `font-mailN' and
  the variable `mail-mode-auto-fill') as suggested by Josef Schugt.
* Implemented the widely-used Emacs bindings:
    - `forward-word' (bound to `M-f'), `backward-word' (bound to `M-b');
    - `capitalize-word' (bount to `M-c');
    - `upcase-word' (bount to `M-u'), `downcase-word' (bount to `M-l');
* Implemented `describe-key-briefly' (bound to `C-h c').
* Implemented `shrink-window' (bound to `M-e').  Contributed by Marius Nita.
* Implemented `back-to-indentation' (bound to `M-m').  Contributed by
  Joseph Boyd <k963@bonbon.net>.
* Implemented `goto-char'.
* Implemented `revisions-kept' and `revisions-delete' variables as
  suggested by Josef Schugt.
* Added `-u rcfile' and `-v variable=value' command line options as
  requested by many users.
* Now multiple functions can be specified on command line with `-f func'.
* Now `quoted-insert' (bound to `C-q') accepts octal digits to
  insert particular characters.
* Cleaned up X Window terminal resize handling.  Low level SIGWINCH signal
  handling has been removed; now Zile relies only on ncurses KEY_RESIZE.
* The behavior of the `auto-font-lock' variable is changed in this version.
  Now the variable applies also to buffer mode changes.
* Fixed two old bugs thanks to Reuben Thomas <rrt@mupsych.org>.
* Patches provided by David A. Capello <dacap@users.sourceforge.net>:
    * implemented `kill-word', `backward-kill-word';
    * fixed bug in undo/redo;
    * fixed bug in highlight;
    * macros support `isearch' facility;
    * shell mode supports highlight for "cat<<EOF\n...\nEOF" stuff.  With
      this, added a new variable: `font-here-document';
    * better shell mode highlight: $identifier, and fixed the \'string\'
      problem;
    * optimized the selection highlight color;
    * added alias `sh-mode' to `shell-script-mode'.
* Integrated some excellent patches from Nicolas Duboc.
* Speed-up improvements.
* Many many other bug fixes.

Changes between Zile 1.6 - Zile 1.6.1

* Added `.m' extension to C Mode (for Objective-C files).
* Fixed window resize bug under xterm (\377 character was inserted).
* Typos and minor bug fixes.

Changes between Zile 1.5.3 - Zile 1.6

* Implemented incremental search (`isearch-forward', `isearch-backward').
* Implemented search & replace (`replace-string', `query-replace').
* Implemented shell commands (`shell-command', `shell-command-on-region').
* Implemented better handling of unexpected errors, with automatic file
  saving on Zile crash.

Changes between Zile 1.5.2 - Zile 1.5.3

* Fixed wrong behavior in novice mode (Font Lock automatically set).
* Fixed a few core dumps.

Changes between Zile 1.5.1 - Zile 1.5.2

* Implemented Shell-script Mode.
* Fixed Solaris compatibility bugs.
* Fixed bugs in C/C++ Font Lock parsing.

Changes between Zile 1.5 - Zile 1.5.1

* Fixed Minibuf filename completion bug.

Changes between Zile 1.4 - Zile 1.5

* Added `switch-to-correlated-buffer' (bound to `F11').
* Added `skip-splash-screen' variable.
* Extented Mini Help functionality; now there are three help pages.
* Reorganized partially the `F1'..`F12' keys.
* Added a reference card (see `etc/refcard.tex' and `etc/refcard.ps').
* User-invisible changes: introduced three new general-purpose libraries:
	- lib/alist.c	Doubly-linked lists
	- lib/astr.c	Dynamically allocated strings
	- lib/htable.c	Hash table functions (was hash.c)
  The new libraries are documented in their man pages (see lib/*.3).
  Starting from this version, the internal structures will be rewritten
  using these libraries.  This will make the code easier to maintain,
  and hopefully fixes some hidden bugs.
* Many bug fixes.

Changes between Zile 1.3 - Zile 1.4

* Added support for `F1'..`F12' keys.
* Implemented the following functions:
	tabify
	untabify
	copy-region-as-kill (bound to `M-w')
* Added support for xterm resizing (SIGWINCH signal).
* Added `text-mode-auto-fill' variable that automatically turns on the
  Auto Fill Mode in Text Mode.
* Now `set-variable' suggests the possibile variable values.
* Added `-f function' command line option for calling internal functions
  at Zile startup.
* Bug fixes.

Changes between Zile 1.2 - Zile 1.3

* Implemented region highlighting.
* Added support for non-English character sets.
* Fixed a security hole that made Zile prone to "symlink attack".

Changes between Zile 1.1 - Zile 1.2

* Fixed some memory leaks thanks to the dmalloc package.
* Added alternative key bindings (`M-h' replaces `C-h') that can be set
  using the `alternative-bindings' variable.
* Now the users with home "/" and users not recorded in the `/etc/passwd'
  file are handled correctly.

Changes between Zile 1.0a5 - Zile 1.1

* Fixed a lot of annoying bugs and updated the documentation.
* Added Auto Fill Mode.
* Added C++ Mode (ISO C++ 14882:1998).
* Updated C Mode keywords to latest C standard (ISO C 9899:1999).
* Implemented `goto-line' (M-g).  Contributed by
  Carsten Leonhardt <leo@debian.org>.
* Added backup directory feature.  Contributed by
  Tim Middelkoop <mtim@arranvale.com>.
* Added `describe-function' (C-h d), `describe-variable' (C-h v) and
  `describe-key' (C-h k).  Now the program is really self-documenting!
* Added `list-bindings' (C-h l b), `list-functions' (C-h l f) and
  `list-registers' (C-h l r).
* Now the search functions remember the last searched value.
* All the help functions bound to `M-h' are now bound to `C-h' like Emacs.
* Added a lot of minor but useful features.

WARNING: the file backup variable `backup-files' has been renamed to

src/astr.c
/* Dynamically allocated strings

   Copyright (c) 2008 Free Software Foundation, Inc.
   Copyright (c) 2001, 2002, 2003, 2004 Sandro Sigala.
   Copyright (c) 2003, 2004, 2005 Reuben Thomas.

   This file is part of GNU Zile.

   GNU Zile is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3, or (at your option)
   any later version.

   GNU Zile is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with GNU Zile; see the file COPYING.  If not, write to the
   Free Software Foundation, Fifth Floor, 51 Franklin Street, Boston,
   MA 02111-1301, USA.  */

#include "config.h"

#include <assert.h>
#include <stdarg.h>
#include <stdlib.h>
#include <string.h>

#include "astr.h"
#include "zile.h"
#include "extern.h"

#define ALLOCATION_CHUNK_SIZE	16

astr
astr_new (void)
{
  astr as;
  as = (astr) XZALLOC (struct astr_s);
  as->maxlen = ALLOCATION_CHUNK_SIZE;
  as->len = 0;
  as->text = (char *) xzalloc (as->maxlen + 1);
  memset (as->text, 0, as->maxlen + 1);
  return as;
}

static void
astr_resize (astr as, size_t reqsize)
{
  assert (as != NULL);
  if (reqsize > as->maxlen)
    {
      as->maxlen = reqsize + ALLOCATION_CHUNK_SIZE;
      as->text = (char *) xrealloc (as->text, as->maxlen + 1);
    }
}

static int
astr_pos (astr as, ptrdiff_t pos)
{
  assert (as != NULL);
  if (pos < 0)
    pos = as->len + pos;
  assert (pos >= 0 && pos <= (int) as->len);
  return pos;
}

char *
astr_char (const astr as, ptrdiff_t pos)
{
  assert (as != NULL);
  pos = astr_pos (as, pos);
  return as->text + pos;
}

void
astr_delete (astr as)
{
  assert (as != NULL);
  free (as->text);
  as->text = NULL;
  free (as);
}

static astr
astr_cpy_x (astr as, const char *s, size_t csize)
{
  astr_resize (as, csize);
  memcpy (as->text, s, csize);
  as->len = csize;
  as->text[csize] = '\0';
  return as;
}

astr
astr_cpy (astr as, const astr src)
{
  assert (src != NULL);
  return astr_cpy_x (as, src->text, src->len);
}

astr
astr_cpy_cstr (astr as, const char *s)
{
  assert (s != NULL);
  return astr_cpy_x (as, s, strlen (s));
}

static astr
astr_cat_x (astr as, const char *s, size_t csize)
{
  astr_resize (as, as->len + csize);
  memcpy (as->text + as->len, s, csize);
  as->len += csize;
  as->text[as->len] = '\0';
  return as;
}

astr
astr_cat (astr as, const astr src)
{
  assert (src != NULL);
  return astr_cat_x (as, src->text, src->len);
}

astr
astr_ncat_cstr (astr as, const char *s, size_t len)
{
  return astr_cat_x (as, s, len);
}

astr
astr_cat_cstr (astr as, const char *s)
{
  return astr_ncat_cstr (as, s, strlen (s));
}

astr
astr_cat_char (astr as, int c)
{
  assert (as != NULL);
  astr_resize (as, as->len + 1);
  as->text[as->len] = (char) c;
  as->text[++as->len] = '\0';
  return as;
}

astr
astr_cat_delete (astr as, const astr src)
{
  assert (src != NULL);
  astr_cat_x (as, src->text, src->len);
  astr_delete (src);
  return as;
}

astr
astr_substr (const astr as, ptrdiff_t pos, size_t size)
{
  assert (as != NULL);
  pos = astr_pos (as, pos);
  assert (pos + size <= as->len);
  return astr_ncat_cstr (astr_new (), astr_char (as, pos), size);
}

static astr
astr_replace_x (astr as, ptrdiff_t pos, size_t size, const char *s,
		size_t csize)
{
  astr tail;

  assert (as != NULL);

  pos = astr_pos (as, pos);
  if (as->len - pos < size)
    size = as->len - pos;
  tail =
    astr_substr (as, pos + (ptrdiff_t) size, astr_len (as) - (pos + size));
  astr_truncate (as, pos);
  astr_ncat_cstr (as, s, csize);
  astr_cat (as, tail);
  astr_delete (tail);

  return as;
}

astr
astr_replace_cstr (astr as, ptrdiff_t pos, size_t size, const char *s)
{
  assert (s != NULL);
  return astr_replace_x (as, pos, size, s, strlen (s));
}

astr
astr_replace_char (astr as, ptrdiff_t pos, int c)
{
  char ch = (char) c;
  return astr_replace_x (as, pos, (size_t) 1, &ch, (size_t) 1);
}

astr
astr_insert_char (astr as, ptrdiff_t pos, int c)
{
  char ch = (char) c;
  return astr_replace_x (as, pos, (size_t) 0, &ch, (size_t) 1);
}

astr
astr_remove (astr as, ptrdiff_t pos, size_t size)
{
  return astr_replace_x (as, pos, size, "", (size_t) 0);
}

/* Don't define in terms of astr_remove, to avoid endless recursion */
astr
astr_truncate (astr as, ptrdiff_t pos)
{
  assert (as != NULL);
  pos = astr_pos (as, pos);
  if ((size_t) pos < as->len)
    {
      as->len = pos;
      as->text[pos] = '\0';
    }
  return as;
}

astr astr_fread(FILE * fp)
{
  int c;
  astr as = astr_new ();

  while ((c = getc (fp)) != EOF)
    astr_cat_char (as, c);

  return as;
}

astr
astr_fgets (FILE * fp)
{
  int c;
  astr as;

  if (feof (fp))
    return NULL;
  as = astr_new ();
  while ((c = fgetc (fp)) != EOF && c != '\n')
    astr_cat_char (as, c);
  return as;
}

astr
astr_vafmt (astr as, const char *fmt, va_list ap)
{
  char *buf;
  xvasprintf (&buf, fmt, ap);
  astr_cat_cstr (as, buf);
  free (buf);
  return as;
}

astr
astr_afmt (astr as, const char *fmt, ...)
{
  va_list ap;
  va_start (ap, fmt);
  astr_vafmt (as, fmt, ap);
  va_end (ap);
  return as;
}

#ifdef TEST

#include <stdio.h>

static void
assert_eq (astr as, const char *s)
{
  if (strcmp(((astr)(as))->text, (s)))
    printf ("test failed: \"%s\" != \"%s\"\n", as->text, s);
}

/*
 * Stubs to make xmalloc_die happy.
 */
char *prog_name = "astr";

void
zile_exit (int doabort GCC_UNUSED)
{
  exit (2);
}


int
main (void)
{
  astr as1, as2, as3;
  int i;
  FILE *fp;

  as1 = astr_new ();
  astr_cpy_cstr (as1, "hello world");
  astr_cat_char (as1, '!');
  assert_eq (as1, "hello world!");

  as3 = astr_substr (as1, 6, 5);
  assert_eq (as3, "world");

  as2 = astr_new ();
  astr_cpy_cstr (as2, "The ");
  astr_cat (as2, as3);
  astr_cat_char (as2, '.');
  assert_eq (as2, "The world.");

  astr_delete (as3);
  as3 = astr_substr (as1, -6, 5);
  assert_eq (as3, "world");

  astr_cpy_cstr (as1, "12345");
  astr_delete (as2);

  astr_cpy_cstr (as1, "12345");
  astr_insert_char (as1, -2, 'x');
  astr_insert_char (as1, -6, 'y');
  astr_insert_char (as1, 7, 'z');
  assert_eq (as1, "y123x45z");

  astr_cpy_cstr (as1, "1234567");
  astr_replace_cstr (as1, -4, 2, "foo");
  assert_eq (as1, "123foo67");

  astr_cpy_cstr (as1, "1234567");
  astr_replace_cstr (as1, 1, 3, "foo");
  assert_eq (as1, "1foo567");

  astr_cpy_cstr (as1, "1234567");
  astr_replace_cstr (as1, -1, 5, "foo");
  assert_eq (as1, "123456foo");

  astr_cpy_cstr (as1, "1234567");
  astr_remove (as1, 4, 10);
  assert_eq (as1, "1234");

  astr_cpy_cstr (as1, "12345");
  as2 = astr_substr (as1, -2, 2);
  assert_eq (as2, "45");

  astr_cpy_cstr (as1, "12345");
  astr_delete (as2);
  as2 = astr_substr (as1, -5, 5);
  assert_eq (as2, "12345");

  astr_cpy_cstr (as1, "1234567");
  astr_replace_cstr (as1, -4, 2, "foo");
  assert_eq (as1, "123foo67");

  astr_cpy_cstr (as1, "1234567");
  astr_replace_cstr (as1, 1, 3, "foo");
  assert_eq (as1, "1foo567");

  astr_cpy_cstr (as1, "1234567");
  astr_replace_cstr (as1, -1, 5, "foo");
  assert_eq (as1, "123456foo");

  astr_cpy_cstr (as1, "");
  astr_afmt (as1, "%s * %d = ", "5", 3);
  astr_afmt (as1, "%d", 15);
  assert_eq (as1, "5 * 3 = 15");
  astr_delete (as1);

  assert (fp = fopen (SRCPATH "/astr.c", "r"));
  as1 = astr_fgets (fp);
  printf ("The first line of astr.c is: \"%s\"\n", astr_cstr (as1));
  assert (fclose (fp) == 0);

  astr_delete (as1);
  astr_delete (as2);
  astr_delete (as3);
  printf ("astr test successful.\n");

  return 0;
}

#endif /* TEST */
